examples/flava/native/train.py

```diff
@@ -0,0 +1,316 @@
+                    )
+                    self.log(
+                        "stats/fwd memory reserved",
+                        torch.cuda.memory_reserved() / 1024**3,
+                    )
+
+                    total_loss = self.calculate_loss(output)
+
+                if self.scaler:
+                    self.scaler.scale(total_loss).backward()
+                    self.scaler.step(optimizer)
+                    self.scaler.update()
+                else:
+                    total_loss.backward()
+                    optimizer.step()
+
+                scheduler.step()
+
+                t1 = time.time()
+                batch_time = t1 - t0
+                batch_size = config.training.batch_size * dist.get_world_size()
+                items_time = batch_size / (t1 - t0)
+
+                t0 = t1
+                self.log("stats/sec per batch", batch_time)
+                self.log("stats/items per sec", items_time)
+
+                total_loss = total_loss.detach()
+                dist.reduce(total_loss, dst=0)
+
+                if self.rank == 0:
+                    norm_total_loss = total_loss.item() / dist.get_world_size()
+
+                    print(
+                        f"epoch: {self.epochs} step {self.steps} loss: {norm_total_loss:.4}"
+                    )
+                    self.log("train/loss", norm_total_loss)
+                    self.log("stats/batch_size", batch_size)
+                self.log(
+                    "stats/max_gpu_allocated_gb",
+                    torch.cuda.max_memory_allocated() / 1024**3,
+                )
+                # TODO implement imagenet eval
+                # TODO implement checkpoint saving
+
+                self.validate()
+
+    def validate(self):
+        if self.steps % self.config.training.validation_steps != 0 or self.steps == 0:
+            return
+
+        model = self.model
+
+        print0("evaluating")
+        model.eval()
+        validation_loader = self.datamodule.val_dataloader()
+        validation_loss = torch.Tensor([0]).to(self.device)
+        for i, data in enumerate(validation_loader):
+            # quick hack to make validation faster
+            if i == 20:
+                break
+            data = self.preprocess_data(data)
+            with torch.no_grad():
+                with torch.cuda.amp.autocast(
+                    dtype=torch.bfloat16
+                ) if self.scaler else nullcontext():
+                    output = model(data)
+                    total_loss = self.calculate_loss(output, validation=True)
+                    validation_loss += total_loss.detach()
+
+        print0(validation_loss)
+        dist.reduce(validation_loss, dst=0)
+        norm_validation_loss = validation_loss.item() / dist.get_world_size()
+        print0(f"step {self.steps} EVAL loss: {norm_validation_loss:.4}")
+        self.log("validation/loss", norm_validation_loss, always_log=True)
```

Comment:
move to compute_loss just like for train?

Reply:
calculate_loss doesn't accumulate the loss, so we need this still